File: 04-SM3-expression-deduction/SM3ExpressionDeduction.cpp
/*
 * @Description: Implement in c++
 * @Autor: zltang
 * @Date: 2022-11-10 15:44:21
 * @LastEditors: Jechin
 * @LastEditTime: 2022-11-10 19:00:25
 */

#include <iostream>
#include <fstream>

#define SS1(A, E, j) ("(((" + A + "<<<12)+" + E + "+(T_" + to_string(j) + "<<<" + to_string(j) + "))<<<7)")
#define SS2(A, SS1Str) ("(" + SS1Str + "^" + "(" + A + "<<<12))")
#define TT1(A, B, C, D, SS2Str, W_, j) ("(" + (j <= 15 ? A + "^" + B + "^" + C : "(" + A + "&" + B + ")" + "|" + "(" + A + "&" + C + ")" + "|" + "(" + B + "&" + C + ")") + "+" + D + "+" + SS2Str + "+" + W_ + to_string(j) + ")")
#define TT2(E, F, G, H, SS1Str, W, j) ("(" + (j <= 15 ? E + "^" + F + "^" + G : "(" + E + "&" + F + ")" + "|" + "(~" + E + "&" + G + ")") + "+" + H + "+" + SS1Str + "+" + W + to_string(j) + ")")
#define D(C) "(" + C +  ")"
#define C(B) (B + "<<<9")
#define B(A) "(" + A  + ")"
#define A(TT1Str) TT1Str
#define H(G) "(" + G + ")"
#define G(F) ("(" + F + "<<<19)")
#define F(E) "(" + E + ")"
#define E(TT2Str) ("((" + TT2Str + ")^((" + TT2Str + ")<<<9)^((" + TT2Str + ")<<<17))")

using namespace std;

int main()
{
    string A = "A_0", B = "B_0", C = "C_0", D = "D_0", E = "E_0", F = "F_0", G = "G_0", H = "H_0", T = "T_0", W = "W_", W_ = "W'_";
    ofstream fileIn;
    fileIn.open("output.txt");
    if(!fileIn.is_open()) return 1;
    for (int j = 0; j < 8; j++)
    {
        string SS1 = SS1(A, E, j);
        string SS2 = SS2(A, SS1);
        string TT1 = TT1(A, B, C, D, SS2, W_, j);
        string TT2 = TT2(E, F, G, H, SS1, W, j);
        D = D(C);
        C = C(B);
        B = B(A);
        A = A(TT1);
        H = H(G);
        G = G(F);
        F = F(E);
        E = E(TT2);
        fileIn << "*************************j = " << j <<  "*****************" << endl;
        fileIn << "A = " << A << endl;
        fileIn << "B = " << B << endl;
        fileIn << "C = " << C << endl;
        fileIn << "D = " << D << endl;
        fileIn << "E = " << E << endl;
        fileIn << "F = " << F << endl;
        fileIn << "G = " << G << endl;
        fileIn << "H = " << H << endl;
        // fileIn << "SS1 = " << SS1 << endl;
        // fileIn << "SS2 = " << SS2 << endl;
        // fileIn << "TT1 = " << TT1 << endl;
        // fileIn << "TT1 = " << TT2 << endl;
        fileIn << endl;
        

    }
    fileIn.close();
    return 0;
}

File: 04-SM3-expression-deduction/SM3ExpressionDeduction.py
'''
Description: 状态更新函数推演
Autor: Jechin
Date: 2022-11-09 16:21:02
LastEditors: Jechin
LastEditTime: 2022-11-10 10:24:55
'''

from tqdm import tqdm
import os

def print_list(var, list):
    with open(f"./VarExpression/{var}.txt", "w") as f:
        for item in list:
            f.write(item)
    print("written in " + var + ".txt")

def SS1(i):
    ans = ["(", "(", "(", f"A_{i}", "<<<12", ")", "+", f"E_{i}", "+", f"(T_{i}<<<{i})", ")", "<<<7", ")"]
    return ans

def extend(list1, list2, index):
    if index > len(list1):
        print("index out of range, extend error")
        exit(0)
    for i in range(len(list2)):
        list1.insert(index + i, list2[i])

    return list1


def SS2(i):
    ans = []
    ans.append("(")
    ans.append(f"SS1_{i}")
    ans.append("+")
    ans.append("(")
    ans.append(f"A_{i}")
    ans.append("<<<12")
    ans.append(")")
    ans.append(")")
    return ans

def TT1(i):
    ans = []
    ans.append("(")
    FF = []
    if i < 16:
        FF += ["(", f"A_{i}", "^", f"B_{i}", "^", f"C_{i}", ")"]
    elif i < 64:
        FF += ["(", "(", f"A_{i}", "&", f"B_{i}", ")", "|", "(", f"A_{i}", "&", f"C_{i}", ")", "|", "(", f"B_{i}", "&", f"C_{i}", ")", ")"]
    else:
        print("TT1 error")
        exit(0)
    ans += FF
    ans += ["+", f"D_{i}", "+", f"SS2_{i}", "+", f"W'_{i}"]

    ans.append(")")
    return ans

def TT2(i):
    ans = []
    ans.append("(")
    GG = []
    if i < 16:
        GG += ["(", f"E_{i}", "^", f"F_{i}", "^", f"G_{i}", ")"]
    elif i < 64:
        GG += ["(", "(", f"E_{i}", "&", f"F_{i}", ")", "|", "(", "~", f"E_{i}", "&", f"G_{i}", ")", ")"]
    else:
        print("TT2 error")
        exit(0)
    ans += GG
    ans += ["+", f"H_{i}", "+", f"SS1_{i}", "+", f"W_{i}"]
    ans.append(")")
    return ans

def D(i):
    if i % 8 == 0:
        return [f"D_{i}"]
    elif os.path.exists(f"./varExpression/D_{i}.txt"):
        with open(f"./varExpression/D_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"C_{i-1}", ")"]

def C(i):
    if i % 8 == 0:
        return [f"C_{i}"]
    elif os.path.exists(f"./varExpression/C_{i}.txt"):
        with open(f"./varExpression/C_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"B_{i-1}", "<<<9", ")"]


def B(i):
    if i % 8 == 0:
        return [f"B_{i}"]
    elif os.path.exists(f"./varExpression/B_{i}.txt"):
        with open(f"./varExpression/B_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"A_{i-1}", ")"]


def A(i):
    if i % 8 == 0:
        return [f"A_{i}"]
    elif os.path.exists(f"./varExpression/A_{i}.txt"):
        with open(f"./varExpression/A_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"TT1_{i-1}", ")"]

def H(i):
    if i % 8 == 0:
        return [f"H_{i}"]
    elif os.path.exists(f"./varExpression/H_{i}.txt"):
        with open(f"./varExpression/H_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"G_{i-1}", ")"]

def G(i):
    if i % 8 == 0:
        return [f"G_{i}"]
    elif os.path.exists(f"./varExpression/G_{i}.txt"):
        with open(f"./varExpression/G_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"F_{i-1}", "<<<19", ")"]

def F(i):
    if i % 8 == 0:
        return [f"F_{i}"]
    elif os.path.exists(f"./varExpression/F_{i}.txt"):
        with open(f"./varExpression/F_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"E_{i-1}", ")"]

def E(i):
    if i % 8 == 0:
        return [f"E_{i}"]
    elif os.path.exists(f"./varExpression/E_{i}.txt"):
        with open(f"./varExpression/E_{i}.txt", "r") as f:
            ans = f.read()
        return [ans]
    else:
        return ["(", f"TT2_{i-1}", "^", "(", f"TT2_{i-1}", "<<<9", ")", "^", "(", f"TT2_{i-1}", "<<<17", ")", ")"]

def Calculate(test):
    iter = 0
    round = 0
    while iter < len(test):
        item = test[iter]
        if item[0:4] == "SS1_":
            SS1_result = SS1(int(item[4:]))
            test.remove(item)
            extend(test, SS1_result, iter)
            round += 1
        elif item[0:4] == "SS2_":
            SS2_result = SS2(int(item[4:]))
            test.remove(item)
            extend(test, SS2_result, iter)
            round += 1
        elif item[0:4] == "TT1_":
            TT1_result = TT1(int(item[4:]))
            test.remove(item)
            extend(test, TT1_result, iter)
            round += 1
        elif item[0:4] == "TT2_":
            TT2_result = TT2(int(item[4:]))
            test.remove(item)
            extend(test, TT2_result, iter)
            round += 1
        elif item[0:2] == "D_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                D_result = D(int(item[2:]))
                test.remove(item)
                extend(test, D_result, iter)
                round += 1
        elif item[0:2] == "C_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                C_result = C(int(item[2:]))
                test.remove(item)
                extend(test, C_result, iter)
                round += 1
        elif item[0:2] == "B_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                B_result = B(int(item[2:]))
                test.remove(item)
                extend(test, B_result, iter)
                round += 1
        elif item[0:2] == "A_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                A_result = A(int(item[2:]))
                test.remove(item)
                extend(test, A_result, iter)
                round += 1
        elif item[0:2] == "H_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                H_result = H(int(item[2:]))
                test.remove(item)
                extend(test, H_result, iter)
                round += 1
        elif item[0:2] == "G_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                G_result = G(int(item[2:]))
                test.remove(item)
                extend(test, G_result, iter)
                round += 1
        elif item[0:2] == "F_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                F_result = F(int(item[2:]))
                test.remove(item)
                extend(test, F_result, iter)
                round += 1
        elif item[0:2] == "E_":
            if int(item[2:]) == 0:
                iter += 1
            else:
                E_result = E(int(item[2:]))
                test.remove(item)
                extend(test, E_result, iter)
                round += 1
        else:
            iter += 1

    return test, round

if __name__ == "__main__":
    # var_list = ["A_1", "B_1", "C_1", "D_1", "E_1", "F_1", "G_1", "H_1"]
    # var_list = ["A_2", "B_2", "C_2", "D_2", "E_2", "F_2", "G_2", "H_2"]
    # var_list = ["A_3", "B_3", "C_3", "D_3", "E_3", "F_3", "G_3", "H_3"]
    # var_list = ["A_4", "B_4", "C_4", "D_4", "E_4", "F_4", "G_4", "H_4"]
    # var_list = ["A_5", "B_5", "C_5", "D_5", "E_5", "F_5", "G_5", "H_5"]
    # var_list = ["A_6", "B_6", "C_6", "D_6", "E_6", "F_6", "G_6", "H_6"]
    # var_list = ["A_7", "B_7", "C_7", "D_7", "E_7", "F_7", "G_7", "H_7"]
    for round in range(8):
        var_list = [f"A_{round}", f"B_{round}", f"C_{round}", f"D_{round}", f"E_{round}", f"F_{round}", f"G_{round}", f"H_{round}"]
        for var in tqdm(var_list):
            test = [var]
            result, _ = Calculate(test)
            print_list(var, result)
